radix-engine/src/kernel/kernel_api.rs

```diff
@@ -76,6 +73,27 @@ pub trait KernelNodeApi {
     ) -> Result<(), RuntimeError>;
 }
 
+/// Internal API for kernel modules.
+/// No kernel state changes are expected as of a result of invoking such APIs, except updating returned references.
+pub trait KernelInternalApi {
+    fn get_module_state(&mut self) -> &mut KernelModuleMixer;
+    fn get_current_depth(&self) -> usize;
+    fn get_current_actor(&self) -> ResolvedActor;
+    fn read_bucket(&mut self, bucket_id: BucketId) -> Option<Resource>;
+    fn read_proof(&mut self, proof_id: BucketId) -> Option<ProofSnapshot>;
```

Comment:
Is this temporary? As buckets/proofs should not be a Kernel abstraction.

Reply:
It's all to support the optional `ExecutionTrace` kernel module. :)

Reply:
Cool, can we add a comment :)